### DISCUSSION TITLE
Reproducibility

### DISCUSSION THREAD

**Discussion by daniele-baccega** [FLAMEGPU/FLAMEGPU2#1119] [Q&A]
Hi,
I have some problems with the reproducibility of the results. I set a seed from the XML file, but running two times the same simulation I obtained different results. I also tried to use the RunPlan to set the seed, but nothing. Why does this happen?

Thank you

**Comment by Robadob** [comment]
Off the top of my head there are two likely causes:

*  The sort used for messaging (particularly spatial and bucket) is not a stable sort, hence the order of messages within bins may vary between runs. If you are doing floating point math with data from messages, this can hence cause minor differences that will compound over time.
* The atomic backed functions for mutating environment macro properties within agent functions are atomic, hence provide no guarantees over order.

There's some notes a colleague previously made here: https://github.com/FLAMEGPU/FLAMEGPU2/issues/417

**Reply by Robadob** [reply]
~~Additional possibility, agents reading spatial messages are now sorted to improve message access performance. This [is not guaranteed](https://thrust.github.io/doc/group__sorting_ga3f47925d80f4970d5730051dba1c5603.html) to be a stable sort either.
Same applies if you use HostAPI agent sort functionality.~~

~~As noted in #417, random state is not attached to agents, so sorting agents can affect determinism.~~

~~The agent sort for messaging can however be disabled via [`AgentDescription::setSortPeriod(0)`](https://github.com/FLAMEGPU/FLAMEGPU2/blob/7a7944d0e3de43882b77e4f9f4dc8fda7b43b6c1/include/flamegpu/model/AgentDescription.h#L310).~~

**Correction:** Explicitly stable sort **is** used for agent sort/HostAPI.
